observer/holder/nodesHolder.go

```diff
@@ -0,0 +1,436 @@
+	lastSyncedNodes        map[uint32]*data.NodeData
+	shardIDs               []uint32
+	availability           data.ObserverDataAvailabilityType
+}
+
+// NewNodesHolder will return a new instance of a nodesHolder
+func NewNodesHolder(syncedNodes []*data.NodeData, fallbackNodes []*data.NodeData, shardIDs []uint32, availability data.ObserverDataAvailabilityType) (*nodesHolder, error) {
+	if len(shardIDs) == 0 {
+		return nil, errEmptyShardIDsList
+	}
+
+	return &nodesHolder{
+		syncedNodes:            syncedNodes,
+		outOfSyncNodes:         make([]*data.NodeData, 0),
+		syncedFallbackNodes:    fallbackNodes,
+		outOfSyncFallbackNodes: make([]*data.NodeData, 0),
+		lastSyncedNodes:        make(map[uint32]*data.NodeData),
+		shardIDs:               shardIDs,
+		availability:           availability,
+	}, nil
+}
+
+// UpdateNodes will update the internal maps based on the provided nodes
+func (nh *nodesHolder) UpdateNodes(nodesWithSyncStatus []*data.NodeData) {
+	if len(nodesWithSyncStatus) == 0 {
+		return
+	}
+	syncedNodes, syncedFallbackNodes, outOfSyncNodes, err := computeSyncedAndOutOfSyncNodes(nodesWithSyncStatus, nh.shardIDs, nh.availability)
+	if err != nil {
+		log.Error("cannot update nodes based on sync state", "error", err)
+		return
+	}
+
+	sameNumOfSynced := len(nh.syncedNodes) == len(syncedNodes)
+	sameNumOfSyncedFallback := len(nh.syncedFallbackNodes) == len(syncedFallbackNodes)
+	if sameNumOfSynced && sameNumOfSyncedFallback && len(outOfSyncNodes) == 0 {
+		nh.printSyncedNodesInShardsUnprotected()
+		// early exit as all the nodes are in sync
+		return
+	}
+
+	syncedNodesMap := nodesSliceToShardedMap(syncedNodes)
+	syncedFallbackNodesMap := nodesSliceToShardedMap(syncedFallbackNodes)
+
+	nh.removeOutOfSyncNodesUnprotected(outOfSyncNodes, syncedNodesMap, syncedFallbackNodesMap)
+	nh.addSyncedNodesUnprotected(syncedNodes, syncedFallbackNodes)
+	nh.printSyncedNodesInShardsUnprotected()
+}
+
+// GetSyncedNodes returns all the synced nodes
+func (nh *nodesHolder) GetSyncedNodes() []*data.NodeData {
+	nh.mut.RLock()
+	defer nh.mut.RUnlock()
+
+	return copyNodes(nh.syncedNodes)
+}
+
+// GetSyncedFallbackNodes returns all the synced fallback nodes
+func (nh *nodesHolder) GetSyncedFallbackNodes() []*data.NodeData {
+	nh.mut.RLock()
+	defer nh.mut.RUnlock()
+
+	return copyNodes(nh.syncedFallbackNodes)
+}
+
+// GetOutOfSyncNodes returns all the out of sync nodes
+func (nh *nodesHolder) GetOutOfSyncNodes() []*data.NodeData {
+	nh.mut.RLock()
+	defer nh.mut.RUnlock()
+
+	return copyNodes(nh.outOfSyncNodes)
+}
+
+// GetOutOfSyncFallbackNodes returns all the out of sync fallback nodes
+func (nh *nodesHolder) GetOutOfSyncFallbackNodes() []*data.NodeData {
+	nh.mut.RLock()
+	defer nh.mut.RUnlock()
+
+	return copyNodes(nh.outOfSyncFallbackNodes)
+}
+
+// GetLastSyncedNodes returns the internal map of the last synced nodes
+func (nh *nodesHolder) GetLastSyncedNodes() map[uint32]*data.NodeData {
+	mapCopy := make(map[uint32]*data.NodeData, 0)
```

Comment:
```suggestion
	mapCopy := make(map[uint32]*data.NodeData, len(nh.lastSyncedNodes))
```
+ move the RLock before

Reply:
refactored